discovery-provider/src/tasks/index_user_listening_history.py

```diff
@@ -0,0 +1,137 @@
+import logging
+import time
+from typing import DefaultDict
+import sqlalchemy as sa
+from src.models.models import UserListeningHistory
+from src.tasks.celery_app import celery
+from src.models import IndexingCheckpoints, Play
+from src.utils.helpers import UPDATE_INDEXING_CHECKPOINTS_QUERY
+
+logger = logging.getLogger(__name__)
+
+USER_LISTENING_HISTORY_TABLE_NAME = "user_listening_history"
+BATCH_SIZE = 100000 # index 100k plays at most at a time
+
+def sort_listening_history(deduped_history_dict, limit = 100):
+    # sorts listening history and places a limit
+    deduped_history = []
+    for track_id, timestamp in deduped_history_dict.items():
+        deduped_history.append({"track_id": track_id, "timestamp": timestamp})
+    deduped_history.sort(key = lambda listen: listen["timestamp"], reverse=True)
+    return deduped_history[:limit]
+
+def _index_user_listening_history(session):
+    # get the last updated id that counted towards user_listening_history
+    # use as lower bound
+    prev_id_checkpoint = (session.query(IndexingCheckpoints.last_checkpoint)
+        .filter(IndexingCheckpoints.tablename == USER_LISTENING_HISTORY_TABLE_NAME)
+    ).scalar()
+
+    if not prev_id_checkpoint:
+        prev_id_checkpoint = 0
+
+    # get new plays since the last checkpoint
+    new_plays = (
+        session.query(Play.id, Play.user_id, Play.play_item_id, Play.created_at)
+        .filter(Play.id > prev_id_checkpoint)
+        .filter(Play.user_id != None)
+        .order_by(sa.asc(Play.id))
+        .limit(BATCH_SIZE)
+    ).all()
+
+    # no update exit early
+    if not new_plays:
+        return
+    new_checkpoint = new_plays[-1].id # get the highest play id
+
+    # get existing user listening history for users with new plays
+    users_with_new_plays = set([new_play.user_id for new_play in new_plays])
+    existing_user_listening_history = (
+        session.query(UserListeningHistory)
+        .filter(UserListeningHistory.user_id.in_(users_with_new_plays))
+        .all()
+    )
+    existing_users = set([user_history.user_id for user_history in existing_user_listening_history])
+
+    # reduce new plays
+    insert_user_listening_history_dict = DefaultDict(list)
+    update_user_listening_history_dict = DefaultDict(list)
+    for new_play in reversed(new_plays):
+        listen_obj ={"track_id": new_play.play_item_id, "timestamp": str(new_play.created_at)}
```

Comment:
I think this would be a fantastic case for us to use a typed class to create this JSON object.

```
class ListenHistory:
  def __init__(self, track_id: Integer, timestamp: String):
    # init
  def to_dict():
    return {"track_id": ... }

```
I also think since this is a complex task, it could be beneficial for us to move this out into a 

src/tasks/listen_history/

subfolder where you could separate these components a bit (like the class be its own file)